src/main/java/at/ac/tuwien/kr/alpha/common/predicates/BaseExternalPredicate.java

```diff
@@ -0,0 +1,42 @@
+package at.ac.tuwien.kr.alpha.common.predicates;
+
+import at.ac.tuwien.kr.alpha.common.terms.ConstantTerm;
+import at.ac.tuwien.kr.alpha.common.terms.Term;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+public abstract class BaseExternalPredicate extends FixedInterpretationPredicate {
+	public BaseExternalPredicate(String name, int arity) {
+		super(name, arity);
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public Set<List<ConstantTerm>> evaluate(List<Term> terms) {
+		if (terms.size() != arity) {
+			throw new IllegalArgumentException(name + " can only be used to test exactly " + arity + " term(s).");
+		}
+
+		final List<ConstantTerm> constants = new ArrayList<>(terms.size());
+		for (int i = 0; i < terms.size(); i++) {
+			if (!(terms.get(i) instanceof ConstantTerm)) {
+				throw new IllegalArgumentException(
+					"Expected only constants as input for " + getPredicateName() + ", but got " +
+						"something else at position " + i + "."
+				);
+			}
+
+			constants.add((ConstantTerm) terms.get(i));
+		}
+
+		try {
+			return test(constants) ? TRUE : FALSE;
```

Comment:
This implies that none of the externals derived from `BaseExternalPredicate` can bind variables. Fine for me.

Reply:
Yes. All classes from the `Predicate` family just return a `boolean`, this should be fine.

Reply:
In the light of this I better understand the class hierarchy. I was wondering why `ExternalBindingMethodPredicate` was not a subtype of `BaseExternalPredicate`. Maybe a more intuitive naming would be to rename `BaseExternalPredicate` to `ExternalNoOutputPredicate` (or `ExternalNoOutputBasePredicate`) and `ExternalBindingMethodPredicate` to `ExternalOutputPredicate`?

Reply:
Is there any particular reason why you suggest to go from "binding" to "output"? I think with this we might introduce even more words that might be juggled around, and that "binding" captures quite well what it outputs.

In the scheme I proposed (last), the class hierarchy branches so that every instantiable class with a name matching `ExternalBinding.*Predicate` will call some external method that will return a set of lists of terms, and everything that just has `External.*Predicate` will call some external method that returns a boolean, which will then be converted into a set of lists of terms.

So the reason why `ExternalBindingMethodPredicate` is not a subtype of `BaseExternalPredicate` is precisely that it will return a set of lists of terms (e.g. bindings), while the others, that are below `BaseExternalPredicate` will not (they just return a boolean).

Reply:
There is some reason, namely that I need to make the grounder care better for what are binding variables and in what order the grounding of a rule body should proceed. I'm thinking of each atom reporting which variables it can bind and which others must be bound already. In case of ordinary atoms this is simple, all positive atoms bind variables, negative ones never bind. But in case of arithmetics and aggregates this is more involved and I'd like to try a better solution than the current one (which is even not according to the standard). Maybe one which allows to rewire the evaluation of atoms depending on things like precision and quantity.

So, "binding" for me has to do with variable binding (located purely in `NaiveGrounder`), which is actually nothing that I expected the external atoms to do. You wrote code, which does variable binding (`getSubstitutions`), so it currently is not so separated as I had in mind and I would like to keep your code.
Would renaming `BaseExternalPredicate` to `ExternalNonBindingBasePredicate` be acceptable for you? If not, then we can also keep it as is.